### DISCUSSION TITLE
Support macOS ICNS icon format

### DISCUSSION THREAD

**Discussion by shoogle** [ImageMagick/ImageMagick#6713] [Contribute]
The ICNS format on macOS is similar to the ICO format on Windows. It can contain multiple image sizes and has support for alpha transparency and various image encodings.

The [Wikipedia article](https://en.wikipedia.org/wiki/Apple_Icon_Image_format) describes the format in more detail and includes pseudocode for reading/writing data chunks.

Advanced features include:

- support for "multiple image states (example: open and closed folders)".
  - Personally I don't care about this feature and I doubt many people make use of it.
- support for `@2x` pixel density for HDPI / "retina" displays.
  - I would like this feature.

My use-case for ICNS is generating icons for a macOS application from SVGs (possibly using [Inkscape](https://inkscape.org/) to covert the SVGs to PNG as an intermediate step).

### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am using the latest version of ImageMagick
- [x] I have searched [open](https://github.com/ImageMagick/ImageMagick/issues) and [closed](https://github.com/ImageMagick/ImageMagick/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

**Comment by RichardBronosky** [comment]
Image Magick is such a natural selection for working with icon files because you often have to operate in batches. I have found a way to work around it though.

```
# convert single .icns file to a several icon_512x512@2x.png files in a .iconset directory
iconutil -c iconset /Applications/Alacritty.app/Contents/Resources/alacritty.icns

# do Image Magick alterations
my_script.sh /Applications/Alacritty.app/Contents/Resources/alacritty.iconset/*

# convert it back
mv /Applications/Alacritty.app/Contents/Resources/alacritty.icns /tmp/
iconutil -c icns /Applications/Alacritty.app/Contents/Resources/alacritty.iconset
mv /Applications/Alacritty.app/Contents/Resources/alacritty.iconset /tmp/
```

**Comment by ewancg** [comment]
> Image Magick is such a natural selection for working with icon files because you often have to operate in batches. I have found a way to work around it though.
> 
> ```
> # convert single .icns file to a several icon_512x512@2x.png files in a .iconset directory
> iconutil -c iconset /Applications/Alacritty.app/Contents/Resources/alacritty.icns
> 
> # do Image Magick alterations
> my_script.sh /Applications/Alacritty.app/Contents/Resources/alacritty.iconset/*
> 
> # convert it back
> mv /Applications/Alacritty.app/Contents/Resources/alacritty.icns /tmp/
> iconutil -c icns /Applications/Alacritty.app/Contents/Resources/alacritty.iconset
> mv /Applications/Alacritty.app/Contents/Resources/alacritty.iconset /tmp/
> ```

The problem with this is that you have to be on macOS to do it. Something could definitely be done with [libicns](https://icns.sourceforge.io)

**Comment by rollingmoai** [comment]
> The problem with this is that you have to be on macOS to do it.

There are at least 2 cross-platform utilities that you can use:
- https://github.com/JackMordaunt/icns
- https://github.com/idesis-gmbh/png2icons (GUI: https://github.com/sprout2000/elephicon)

Additionally, the latest development version of GIMP (`2.99.14`) has added support for the ICNS format.

**Comment by frozenpandaman** [comment]
Would love to see this supported someday…!

**Comment by frozenpandaman** [comment]
@eclairevoyant Not sure what's up with the 👎 literally 60 seconds after I posted this comment or how you're involved with the project but impressive reaction speed!

Edit: Oh no, the horror, six people getting an email. My point was to bump it, yes, intentionally! I personally unsubscribe from issues in which I have no stake or don't want to engage in any sort of further discussion – but I guess people operate all sorts of ways, though, so perhaps that's not the case for you or others (?) and possibly a very select few people got an email they didn't want. As happens every day in [current year], I think. Alas.

**Comment by hhyyrylainen** [comment]
You do realize that everyone who is watching this issue (including me) will get an email about each new comment on this? Which is pretty annoying if people post comments that are just basically saying +1 to the importance of this issue. That is something that can be achieved by thumbs upping the issue itself to avoid annoying people with pointless emails much more elegantly. Sorry in advance to everyone who gets an email about this comment.
